File: java/src/main/java/sk/epostak/sdk/resources/WebhookQueueResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.AckResponse;
import sk.epostak.sdk.models.BatchAckResponse;
import sk.epostak.sdk.models.WebhookQueueAllResponse;
import sk.epostak.sdk.models.WebhookQueueResponse;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Webhook pull queue for polling-based event consumption.
 * <p>
 * Alternative to push webhooks -- pull events from a queue at your own pace.
 * Events remain in the queue until explicitly acknowledged.
 * <p>
 * Access via {@code client.webhooks().queue()}.
 *
 * <pre>{@code
 * // Poll for new events and acknowledge them
 * WebhookQueueResponse response = client.webhooks().queue().pull(50, null);
 * List<String> ids = response.items().stream()
 *     .map(WebhookQueueResponse.WebhookQueueItem::id)
 *     .toList();
 * client.webhooks().queue().batchAck(ids);
 * }</pre>
 */
public final class WebhookQueueResource {

    private final HttpClient http;

    /**
     * Creates a new webhook queue resource.
     *
     * @param http the HTTP client used for API communication
     */
    WebhookQueueResource(HttpClient http) {
        this.http = http;
    }

    /**
     * Fetch pending events from the queue.
     *
     * @param limit     max items to return (1-100, default 20), or {@code null} for default
     * @param eventType optional event type filter (e.g. {@code "document.received"}), or {@code null} for all
     * @return the queue response with items and a {@code hasMore} flag
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public WebhookQueueResponse pull(Integer limit, String eventType) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("limit", limit);
        params.put("event_type", eventType);
        return http.get("/webhook-queue" + HttpClient.buildQuery(params), WebhookQueueResponse.class);
    }

    /**
     * Fetch pending events from the queue with default parameters.
     *
     * @return the queue response (limit 20, all event types)
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public WebhookQueueResponse pull() {
        return pull(null, null);
    }

    /**
     * Acknowledge (remove) a single event from the queue.
     *
     * @param eventId the event UUID to acknowledge
     * @return acknowledgement confirmation with {@code acknowledged = true}
     * @throws sk.epostak.sdk.EPostakException if the event is not found or the request fails
     */
    public AckResponse ack(String eventId) {
        return http.delete("/webhook-queue/" + HttpClient.encode(eventId), AckResponse.class);
    }

    /**
     * Batch acknowledge (remove) multiple events from the queue.
     *
     * @param eventIds list of event UUIDs to acknowledge
     * @return response containing the count of acknowledged events
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public BatchAckResponse batchAck(List<String> eventIds) {
        Map<String, Object> body = Map.of("event_ids", eventIds);
        return http.post("/webhook-queue/batch-ack", body, BatchAckResponse.class);
    }

    /**
     * Fetch events across all firms. Requires an integrator key ({@code sk_int_*}).
     *
     * @param limit max items to return (1-500, default 100), or {@code null} for default
     * @param since only return events created after this ISO 8601 timestamp, or {@code null}
     * @return the cross-firm queue response with events and count
     * @throws sk.epostak.sdk.EPostakException if the request fails or the key is not an integrator key
     */
    public WebhookQueueAllResponse pullAll(Integer limit, String since) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("limit", limit);
        params.put("since", since);
        return http.get("/webhook-queue/all" + HttpClient.buildQuery(params), WebhookQueueAllResponse.class);
    }

    /**
     * Fetch events across all firms with default parameters.
     *
     * @return the cross-firm queue response (limit 100)
     * @throws sk.epostak.sdk.EPostakException if the request fails or the key is not an integrator key
     */
    public WebhookQueueAllResponse pullAll() {
        return pullAll(null, null);
    }

    /**
     * Batch acknowledge events across all firms. Requires an integrator key ({@code sk_int_*}).
     *
     * @param eventIds list of event UUIDs to acknowledge
     * @return the response containing the count of acknowledged events
     * @throws sk.epostak.sdk.EPostakException if the request fails or the key is not an integrator key
     */
    public BatchAckAllResponse batchAckAll(List<String> eventIds) {
        Map<String, Object> body = Map.of("event_ids", eventIds);
        return http.post("/webhook-queue/all/batch-ack", body, BatchAckAllResponse.class);
    }

    /**
     * Response from cross-firm batch acknowledge.
     *
     * @param acknowledged the number of events that were successfully acknowledged
     */
    public record BatchAckAllResponse(int acknowledged) {}
}

File: dotnet/src/EPostak/Resources/WebhookQueueResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Polling-based webhook queue for consuming events without exposing a public endpoint.
/// Events remain in the queue until explicitly acknowledged. This is an alternative to
/// push webhooks for systems behind firewalls or without stable public URLs.
/// </summary>
public sealed class WebhookQueueResource
{
    private readonly HttpRequestor _http;

    internal WebhookQueueResource(HttpRequestor http) => _http = http;

    /// <summary>
    /// Pull pending events from the webhook queue. Events remain in the queue until
    /// acknowledged via <see cref="AckAsync"/> or <see cref="BatchAckAsync"/>.
    /// Call this periodically to consume new events.
    /// </summary>
    /// <param name="params">Optional filters: max items to return (1-100) and event type filter.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>List of pending queue items and a flag indicating whether more items are available.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Webhooks.Queue.PullAsync(new WebhookQueueParams
    /// {
    ///     Limit = 10,
    ///     EventType = WebhookEvents.DocumentReceived
    /// });
    /// foreach (var item in response.Items)
    /// {
    ///     Console.WriteLine($"Event {item.EventId}: {item.Event}");
    ///     // Process the event...
    ///     await client.Webhooks.Queue.AckAsync(item.EventId);
    /// }
    /// </code>
    /// </example>
    public Task<WebhookQueueResponse> PullAsync(WebhookQueueParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("limit", @params?.Limit?.ToString()),
            ("event_type", @params?.EventType));
        return _http.RequestAsync<WebhookQueueResponse>(HttpMethod.Get, $"/webhook-queue{qs}", ct);
    }

    /// <summary>
    /// Acknowledge and remove a single event from the queue.
    /// Call this after you have successfully processed the event.
    /// </summary>
    /// <param name="eventId">The queue event UUID to acknowledge.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Acknowledgement confirmation with <c>Acknowledged = true</c>.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Webhooks.Queue.AckAsync("evt_abc123");
    /// Console.WriteLine(result.Acknowledged); // true
    /// </code>
    /// </example>
    public Task<AckResponse> AckAsync(string eventId, CancellationToken ct = default)
        => _http.RequestAsync<AckResponse>(HttpMethod.Delete, $"/webhook-queue/{Uri.EscapeDataString(eventId)}", ct);

    /// <summary>
    /// Acknowledge and remove multiple events from the queue in a single request.
    /// More efficient than calling <see cref="AckAsync"/> individually for each event.
    /// </summary>
    /// <param name="eventIds">Collection of queue event UUIDs to acknowledge.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Object with the count of acknowledged events.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Webhooks.Queue.PullAsync();
    /// // Process all events...
    /// var ids = response.Items.Select(i => i.EventId);
    /// var result = await client.Webhooks.Queue.BatchAckAsync(ids);
    /// Console.WriteLine($"Acknowledged {result.Acknowledged} events");
    /// </code>
    /// </example>
    public Task<BatchAckResponse> BatchAckAsync(IEnumerable<string> eventIds, CancellationToken ct = default)
        => _http.RequestAsync<BatchAckResponse>(HttpMethod.Post, "/webhook-queue/batch-ack", new { event_ids = eventIds }, ct);

    /// <summary>
    /// Pull pending events across all firms managed by the integrator.
    /// Only available with integrator API keys (<c>sk_int_*</c>). Each event includes
    /// the <c>FirmId</c> so you can route it to the correct client.
    /// </summary>
    /// <param name="params">Optional filters: max items (1-500) and ISO 8601 timestamp cutoff.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Cross-firm events with firm IDs and total count.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Webhooks.Queue.PullAllAsync(new WebhookQueueAllParams
    /// {
    ///     Limit = 100,
    ///     Since = "2026-04-01T00:00:00Z"
    /// });
    /// foreach (var evt in response.Items)
    ///     Console.WriteLine($"[Firm {evt.FirmId}] {evt.Event}: {evt.EventId}");
    /// </code>
    /// </example>
    public Task<WebhookQueueAllResponse> PullAllAsync(WebhookQueueAllParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("limit", @params?.Limit?.ToString()),
            ("since", @params?.Since));
        return _http.RequestAsync<WebhookQueueAllResponse>(HttpMethod.Get, $"/webhook-queue/all{qs}", ct);
    }

    /// <summary>
    /// Acknowledge events across all firms in a single batch request.
    /// Only available with integrator API keys (<c>sk_int_*</c>).
    /// </summary>
    /// <param name="eventIds">Collection of cross-firm event UUIDs to acknowledge.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The count of successfully acknowledged events.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Webhooks.Queue.PullAllAsync();
    /// // Process events...
    /// var ids = response.Items.Select(e => e.EventId);
    /// var result = await client.Webhooks.Queue.BatchAckAllAsync(ids);
    /// Console.WriteLine($"Acknowledged {result.Acknowledged} events");
    /// </code>
    /// </example>
    public Task<BatchAckAllResponse> BatchAckAllAsync(IEnumerable<string> eventIds, CancellationToken ct = default)
        => _http.RequestAsync<BatchAckAllResponse>(HttpMethod.Post, "/webhook-queue/all/batch-ack", new { event_ids = eventIds }, ct);
}
